### PR TITLE
[Issue - #40] Add sqlc queries for auth and owned resources

### LINKED ISSUE #40: [Backend Migration 08] Add sqlc queries for auth and owned resources
Parent spec: #18.
Related epics: #21, #22, #23, #24, and #25.
Spec source: `docs/specs/18-go-backend-replacement/spec.md`.

Graphify note:
- Use the current graph output to confirm impacted endpoints, flows, and business logic before implementing changes.
- Refresh the graphify artifacts before handing this ticket in or before starting another migration ticket.

Add typed sqlc query coverage for the Go backend after migrations exist.

Acceptance criteria:
- sqlc generates typed Go accessors for users and auth token/session tables.
- sqlc generates typed Go accessors for accounts and categories CRUD/bulk behavior.
- sqlc generates typed Go accessors for transactions CRUD/list/bulk behavior.
- sqlc generates typed Go accessors or query helpers for summary aggregation.
- Queries are written to work with the authenticated DB session/RLS model.
- Generated query code is not hand edited.

Verification:
- `cd backend && go test ./...`
- sqlc generation command runs successfully.

### PR DESCRIPTION
**Pull request by GonzaloSecades**
Closes #40.

## What

The typed query layer: six hand-written SQL files in `backend/internal/db/queries/` (~35 named queries) and the sqlc-generated Go code in `backend/internal/db/gen/` (committed, never hand-edited). Plus `sql_package: "pgx/v5"` in sqlc.yaml (the pool is pgxpool), a live-gated round-trip test, README sqlc docs, and post-migration-improvements entry 0006.

No handlers, no JWT, no middleware — this is the vocabulary later tickets (#41/#43/#44+) speak.

## Query catalog

- **users.sql** — create (23505 → handler 409), get by email/id, atomic MarkUserEmailVerified, UpdateUserPassword.
- **auth_tokens.sql** — per #41's recorded decisions: create/consume for verification + reset tokens where consume is a single atomic `UPDATE ... WHERE used_at IS NULL AND expires_at > now() RETURNING user_id` (concurrent double-submit safe); invalidate-prior-on-issue; CountRecent* for rate limits; refresh create, read-only get-valid, atomic `ConsumeRefreshToken` for rotation (exactly one concurrent refresh wins), revoke, revoke-all.
- **accounts.sql / categories.sql** — CRUD with `user_id` ownership predicates, `ANY(array)` bulk delete RETURNING only deleted owned ids (fixtures: silent ignore), ORDER BY name for list determinism.
- **transactions.sql** — every read/update/delete owns rows through the accounts join (mirroring the RLS policy); list matches the fixture item shape (joined account/category names with explicit category ownership, inclusive date bounds, optional accountId via narg, date DESC); bulk create is a single `INSERT ... SELECT` from one structured `jsonb` payload via `jsonb_to_recordset` — no parallel arrays to keep in sync, JSON nulls land as SQL NULLs, and an invalid row fails the batch atomically.
- **summary.sql** — period totals (SUM CASE, overflow-safe `ABS(amount::bigint)`, COALESCE 0), category spending (owned inner join excludes uncategorized, ORDER BY value DESC), daily totals grouped by calendar day (`t.date::date`); top-3+"Other" and zero-filling stay in Go like legacy JS.

**Invariant:** ownership predicates in every owned-resource read/update/delete even though FORCE RLS is live — RLS stays the backstop, not the mechanism. The transaction INSERTs are the documented exception: handler pre-validation (legacy friendly 400/404) + RLS `WITH CHECK`, because an ownership join inside the bulk INSERT would silently drop unowned rows instead of failing loudly and atomically.

## Review history

- **[high, parity review, pre-PR]** BulkCreateTransactions couldn't represent NULL notes/category_id (sqlc emits non-nullable `[]string` for array params) — the CSV-import path would have FK-failed the whole batch. Interim `''`-as-NULL sentinel fix, superseded in round 1 by the `jsonb_to_recordset` shape above.
- **Round 1 (4 high from Gonzalo, 2 medium from Copilot), commit f601694:** atomic `ConsumeRefreshToken` rotation primitive (+ concurrent two-goroutine live test), jsonb bulk create replacing parallel unnest arrays (kills silent length-mismatch truncation), daily totals grouped by `t.date::date`, `ABS(t.amount::bigint)` overflow fix (+ boundary test at -2147483648), explicit category-join ownership predicates in ListTransactions and GetCategorySpending.
- **Round 2 (1 medium, 1 low from Copilot), commit 3586058:** README documents the insert ownership model accurately; stale sentinel reference removed from the test header.
- Divergences from the briefing where legacy won: `GetTransaction` returns own columns only (legacy GET /:id has no joined names).

## Post-migration improvements

Entry **0006** (flagged during implementation, recorded per the new rule): `transactions.amount` is 32-bit — caps a single transaction near ±2.1M ARS milliunits, a realistic figure in ARS. Ported as-is for parity; proposed bigint widening post-#27. High priority alongside 0005.

## Verification

- `sqlc generate` clean (pinned v1.31.1); `sqlc diff` proves gen/ matches queries (nothing hand-edited).
- `go vet ./...` clean; `go test ./...` green without a database (live tests skip).
- Live against PostgreSQL 17.7 (WSL): full `internal/db` suite PASS including `TestSqlcQueries_LiveDatabase` — user→account→category→transactions round trip under `set_config('app.user_id', ...)` (FORCE RLS active), ListTransactions joined shape + inclusive bounds + DESC, BulkCreateTransactions with JSON nulls + populated row + atomic failure of a batch containing an invalid row, BulkDeleteTransactions ignoring an unowned id (victim row survives), same-day different-times daily grouping, boundary amount -2147483648 through the summary queries, atomic reset-token consume (second attempt ErrNoRows), RevokeAllUserRefreshTokens — plus `TestConsumeRefreshToken_Concurrent` (two goroutines, exactly one winner). Rollback/cleanup-clean (0 leftover rows).
- graphify refreshed.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: backend/internal/db/queries/auth_tokens.sql [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+-- name: GetRefreshTokenByHash :one
+-- Only returns a row when the token is currently valid (not revoked, not
+-- expired); filtering happens in SQL rather than leaving expiry/revocation
+-- checks to the caller.
+SELECT *
+FROM refresh_tokens
+WHERE token_hash = sqlc.arg(token_hash)
+  AND revoked_at IS NULL
+  AND expires_at > now();
```
Comment (GonzaloSecades) [inline backend/internal/db/queries/auth_tokens.sql:82]: [high] `GetRefreshTokenByHash` is a read-only validity check, so a refresh handler must select, then revoke, then create a replacement. Two concurrent requests can both read the token as valid before either revocation commits and can mint two valid successor sessions. Provide an atomic consume query (`UPDATE ... SET revoked_at = now() WHERE token_hash = $1 AND revoked_at IS NULL AND expires_at > now() RETURNING user_id/id`) and perform successor creation in the same transaction. Add a concurrent double-refresh test proving only one consume succeeds.

File: backend/internal/db/queries/transactions.sql [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    currencies.val
+FROM unnest(sqlc.arg(ids)::text[]) WITH ORDINALITY AS ids(val, ord)
+JOIN unnest(sqlc.arg(amounts)::integer[]) WITH ORDINALITY AS amounts(val, ord) USING (ord)
+JOIN unnest(sqlc.arg(payees)::text[]) WITH ORDINALITY AS payees(val, ord) USING (ord)
+JOIN unnest(sqlc.arg(notes)::text[]) WITH ORDINALITY AS notes(val, ord) USING (ord)
+JOIN unnest(sqlc.arg(dates)::timestamp[]) WITH ORDINALITY AS dates(val, ord) USING (ord)
+JOIN unnest(sqlc.arg(account_ids)::text[]) WITH ORDINALITY AS account_ids(val, ord) USING (ord)
+JOIN unnest(sqlc.arg(category_ids)::text[]) WITH ORDINALITY AS category_ids(val, ord) USING (ord)
+JOIN unnest(sqlc.arg(currencies)::text[]) WITH ORDINALITY AS currencies(val, ord) USING (ord)
+RETURNING *;
```
Comment (GonzaloSecades) [inline backend/internal/db/queries/transactions.sql]: [high] The ordinality joins silently truncate to the shortest parameter array. If any handler mapping produces arrays of different lengths, the query succeeds while dropping financial records, and `RETURNING` alone does not distinguish that from a valid shorter import. Make cardinality equality a database-enforced precondition or pass one structured JSON/composite array, and add a test showing mismatched lengths fail atomically rather than partially inserting.

File: backend/internal/db/queries/summary.sql [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+SELECT
+    t.date AS date,
+    COALESCE(SUM(CASE WHEN t.amount >= 0 THEN t.amount ELSE 0 END), 0)::bigint AS income,
+    COALESCE(SUM(CASE WHEN t.amount < 0 THEN ABS(t.amount) ELSE 0 END), 0)::bigint AS expenses
+FROM transactions t
+JOIN accounts a ON a.id = t.account_id AND a.user_id = sqlc.arg(user_id)
+WHERE t.date >= sqlc.arg(start_date)
+  AND t.date <= sqlc.arg(end_date)
+  AND (sqlc.narg(account_id)::text IS NULL OR t.account_id = sqlc.narg(account_id))
+GROUP BY t.date
```
Comment (GonzaloSecades) [inline backend/internal/db/queries/summary.sql]: [high] This groups by the full timestamp, not by calendar day. Two transactions on the same date at different times produce separate rows; the planned legacy-style zero-fill maps rows by `yyyy-MM-dd`, so one aggregate overwrites the other and the chart silently under-reports that day. Aggregate on an explicit day expression (`t.date::date` or the agreed timezone-aware equivalent) and cover two different timestamps on the same day.

File: backend/internal/db/queries/summary.sql [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+-- computes percentage change, top-3 + "Other" category bucketing, and
+-- missing-day zero-filling in Go, matching legacy JS.
+
+-- name: GetPeriodTotals :one
+-- income/expenses/remaining are SUM(...)::bigint, coalesced to 0 so an empty
+-- period returns zeros instead of NULL (legacy JS does `value || 0`).
+SELECT
+    COALESCE(SUM(CASE WHEN t.amount >= 0 THEN t.amount ELSE 0 END), 0)::bigint AS income,
+    COALESCE(SUM(CASE WHEN t.amount < 0 THEN ABS(t.amount) ELSE 0 END), 0)::bigint AS expenses,
+    COALESCE(SUM(t.amount), 0)::bigint AS remaining
```
Comment (GonzaloSecades) [inline backend/internal/db/queries/summary.sql:15]: [high] `ABS(t.amount)` is evaluated as `integer` before the surrounding `SUM`/cast. PostgreSQL raises `integer out of range` for the valid column value `-2147483648`, so one boundary transaction can make the financial summary fail. Cast before taking the absolute value (`ABS(t.amount::bigint)`) here and in category/daily totals, with a boundary-value live test.

**Review by GonzaloSecades — COMMENTED** [review]
The query layer is well scoped and the generated-code workflow is clean, but I found four blocking correctness/security issues beyond Copilot's ownership comments. They affect refresh-token single use, bulk financial-record integrity, daily aggregation, and an integer boundary that can make summary queries fail. Please address the inline findings and add focused live tests for each before merging.

File: backend/internal/db/queries/auth_tokens.sql (hunk 1 above)
Comment (GonzaloSecades) [reply backend/internal/db/queries/auth_tokens.sql:82]: [high, fixed] f601694 — added `ConsumeRefreshToken`: single-statement `UPDATE ... SET revoked_at = now() WHERE token_hash = $1 AND revoked_at IS NULL AND expires_at > now() RETURNING id, user_id`. Exactly one of any number of concurrent refreshes wins; the winner creates the successor in the same transaction (#43/#41 wiring). `GetRefreshTokenByHash` is kept but documented as read-only-not-for-rotation. Proven by the new `TestConsumeRefreshToken_Concurrent`: two goroutines on separate pool connections race on committed rows — exactly 1 success, 1 `ErrNoRows`, PASS live.

File: backend/internal/db/queries/transactions.sql (hunk 2 above)
Comment (GonzaloSecades) [reply backend/internal/db/queries/transactions.sql]: [high, fixed] f601694 — took your structured-parameter option, and it turned out cleaner than a cardinality guard: `BulkCreateTransactions` now takes a single `jsonb` array via `jsonb_to_recordset` (sqlc's analyzer accepts it, generating `payload []byte`). Length mismatch is now structurally impossible — there are no parallel arrays — and JSON nulls land as SQL NULLs natively, which also retires the previous ''-sentinel workaround. Atomicity is tested live: a batch with an id-less row fails whole (NOT NULL violation), and the valid sibling row is proven absent afterwards.

File: backend/internal/db/queries/summary.sql (hunk 3 above)
Comment (GonzaloSecades) [reply backend/internal/db/queries/summary.sql]: [high, fixed] f601694 — `GetDailyTotals` now selects and groups by `t.date::date`. You're right that grouping by raw timestamp only worked by accident of every current write path storing UTC midnight; one future non-midnight write would have silently split a day and the yyyy-MM-dd zero-fill would drop rows. Live test added: two transactions on the same day at 00:00 and 15:30 aggregate into exactly one row with summed expenses.

File: backend/internal/db/queries/summary.sql (hunk 4 above)
Comment (GonzaloSecades) [reply backend/internal/db/queries/summary.sql:15]: [high, fixed] f601694 — `ABS(t.amount::bigint)` in all three summary queries (period totals, category spending including its ORDER BY, daily totals). Verified the failure first: `ABS(-2147483648)` on integer raises `integer out of range` exactly as you said. Boundary live test added: a transaction with amount -2147483648 flows through GetDailyTotals and GetPeriodTotals correctly (expenses = 2147484648 with a second -1000 row).

File: backend/internal/db/queries/transactions.sql [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+    c.name AS category,
+    t.category_id,
+    t.payee,
+    t.amount,
+    t.notes,
+    a.name AS account,
+    t.account_id
+FROM transactions t
+JOIN accounts a ON a.id = t.account_id AND a.user_id = sqlc.arg(user_id)
+LEFT JOIN categories c ON c.id = t.category_id
```
[the remark this answers is not in the document]
Comment (GonzaloSecades) [reply backend/internal/db/queries/transactions.sql]: [medium, fixed] f601694 — `ListTransactions`' category LEFT JOIN now carries `AND c.user_id = sqlc.arg(user_id)`. For well-formed data the result is identical (writes validate category ownership and RLS backstops), but the invariant is now explicit in the query: a cross-user category_id yields a NULL name instead of leaking a foreign category name.

File: backend/internal/db/queries/summary.sql [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+-- Expense rows only (amount < 0), inner-joined to categories (legacy
+-- innerJoin excludes uncategorized transactions from the breakdown).
+-- Top-3 + "Other" bucketing stays in Go, as in legacy JS.
+SELECT
+    c.name AS name,
+    COALESCE(SUM(ABS(t.amount)), 0)::bigint AS value
+FROM transactions t
+JOIN accounts a ON a.id = t.account_id AND a.user_id = sqlc.arg(user_id)
+JOIN categories c ON c.id = t.category_id
+WHERE t.amount < 0
```
[the remark this answers is not in the document]
Comment (GonzaloSecades) [reply backend/internal/db/queries/summary.sql:32]: [medium, fixed] f601694 — `GetCategorySpending`'s categories join now carries `AND c.user_id = sqlc.arg(user_id)`, same reasoning as the list query: explicit ownership on every joined owned table, so no cross-user category-name leakage even under elevated-privilege execution.

**Comment by GonzaloSecades** [comment]
## Review cycle — iteration 1

**Addressed (6/6)** — Copilot 2 medium + Gonzalo 4 high, all on point, all fixed in f601694:

1. [high] Refresh rotation race → new atomic `ConsumeRefreshToken` (single UPDATE, one winner); proven by a genuinely concurrent two-goroutine live test on committed rows (1 success, 1 ErrNoRows).
2. [high] Bulk-create array truncation → replaced parallel unnest arrays with a single structured `jsonb` payload (`jsonb_to_recordset`): mismatch structurally impossible, JSON nulls native (retires the '' sentinel), invalid row fails the batch atomically (tested).
3. [high] Daily totals grouped by raw timestamp → now `t.date::date`; two same-day timestamps aggregate into one row (tested).
4. [high] `ABS(integer)` overflow at -2147483648 → `ABS(t.amount::bigint)` in all three summary queries; boundary transaction flows through summary correctly (tested).
5. [medium] `ListTransactions` category join → explicit `c.user_id` ownership predicate.
6. [medium] `GetCategorySpending` category join → same.

**Rebutted:** none — every finding survived scrutiny; two (refresh race, ABS overflow) were independently verified as real failures before fixing.

**Verification:** `sqlc generate` + `go vet` clean; full live suite PASS against PostgreSQL 17.7 including the three new tests (concurrent consume, same-day aggregation + int32-boundary summary, bulk atomic failure); DB left clean.

Copilot re-reviews on this push. 1 of 3 automated iterations used.

File: backend/README.md [hunk 7]
```diff
@@ -120,24 +120,67 @@ touched by user-scoped queries.
...
+| `auth_tokens.sql` | `email_verification_tokens`, `password_reset_tokens` (atomic one-time consume, invalidate-prior, rate-limit counts), `refresh_tokens` (create, get-valid, revoke, revoke-all) |
+| `accounts.sql` | `accounts` CRUD + bulk delete |
+| `categories.sql` | `categories` CRUD + bulk delete (mirrors `accounts.sql`) |
+| `transactions.sql` | `transactions` CRUD, list (joined account/category names), bulk create, bulk delete — every query is scoped through the owning account |
+| `summary.sql` | period totals, category spending, daily totals aggregates |
+
+Every owned-resource query (accounts/categories/transactions) carries an
+explicit ownership predicate (`user_id = $N`, or a join/EXISTS through the
+owning account for transactions) even though RLS also enforces it — RLS is
+the backstop, not the mechanism (see RLS above).
```
[the remark this answers is not in the document]
Comment (GonzaloSecades) [reply backend/README.md]: [medium, fixed] 3586058 — README adjusted to describe the actual design rather than adding predicates to the INSERTs: reads/updates/deletes carry explicit ownership; the transaction INSERTs deliberately rely on handler pre-validation (legacy friendly 400/404) + RLS WITH CHECK. Adding an ownership join inside the bulk INSERT was considered and rejected: it would *silently drop* unowned rows — a partial import — whereas WITH CHECK fails loudly and atomically, the safer failure mode for financial data. That rationale is now in the README paragraph.

File: backend/internal/db/queries_live_test.go [hunk 8]
```diff
@@ -0,0 +1,560 @@
...
+	"github.com/jackc/pgx/v5"
+	"github.com/jackc/pgx/v5/pgtype"
+)
+
+// TestSqlcQueries_LiveDatabase exercises a representative path through the
+// sqlc-generated query code (backend/internal/db/gen), not every query:
+// create user -> account -> category -> transactions; ListTransactions
+// (joined names, inclusive date filtering, DESC order); BulkCreateTransactions
+// (single round trip, '' NULL sentinel for notes/category_id landing as SQL
+// NULL); BulkDeleteTransactions
```
[the remark this answers is not in the document]
Comment (GonzaloSecades) [reply backend/internal/db/queries_live_test.go]: [low, fixed] 3586058 — test header comment updated: it now describes the jsonb payload with JSON nulls and atomic batch failure; the '' sentinel reference was a leftover from the retired round-1 approach.

**Comment by GonzaloSecades** [comment]
## Review cycle — rounds 1 and 2

**Round 1 (6 findings — 2 Copilot medium, 4 Gonzalo high), all fixed in f601694:**
1. [high] Refresh rotation race → atomic `ConsumeRefreshToken` (single UPDATE ... RETURNING), proven by a concurrent two-goroutine live test: exactly 1 winner.
2. [high] Bulk-create array cardinality truncation → single structured `jsonb` payload via `jsonb_to_recordset`; mismatch structurally impossible, JSON nulls native, invalid row fails the batch atomically (tested).
3. [high] Daily totals grouped by raw timestamp → now `t.date::date`; same-day different-times test added.
4. [high] `ABS(integer)` overflow at -2147483648 → `ABS(t.amount::bigint)` in all three summary queries; boundary live test added.
5. [medium] `ListTransactions` category join ownership predicate added.
6. [medium] `GetCategorySpending` category join ownership predicate added.

**Round 2 (2 findings), fixed in 3586058:**
7. [medium] README overclaimed ownership predicates on INSERTs → paragraph now documents the deliberate insert model (handler pre-validation + RLS WITH CHECK; an ownership join in the bulk INSERT would silently drop rows instead of failing atomically).
8. [low] Stale '' sentinel reference in the test header comment → updated for the jsonb payload.

**Verification:** `go vet` clean; full suite green; all live tests PASS against PostgreSQL 17.7 including the new concurrent-consume, boundary-amount, same-day-grouping, and batch-atomicity tests; leaked rows from a mid-development cleanup bug removed (users table back to 0).

Copilot re-reviews on this push. 2 of 3 automated iterations used.

File: backend/README.md [hunk 9]
```diff
@@ -120,24 +120,74 @@ touched by user-scoped queries.
...
+`pgx.ErrNoRows`.
+
+`BulkCreateTransactions` inserts every row in one round trip from a single
+structured `jsonb` parameter (`jsonb_to_recordset`): the caller marshals the
+rows as one JSON array (fields `id`, `amount`, `payee`, `notes`, `date`,
+`account_id`, `category_id`, `currency`; dates in UTC — the `timestamp`
+cast drops any zone suffix). One parameter makes per-row integrity
+structural — there are no parallel arrays whose lengths can drift, JSON
+nulls land as SQL NULLs, and a batch containing an invalid row fails
+atomically (single INSERT statement). Refresh-token rotation must use the
```
[the remark this answers is not in the document]
Comment (GonzaloSecades) [reply backend/README.md:170]: [low, fixed] PR description updated — the bulk-create bullet now describes the shipped `jsonb_to_recordset` single-payload shape (the `unnest WITH ORDINALITY` text was the superseded round-0 design), and a new Review history section traces the evolution (parity review → round 1 jsonb rewrite → round 2 docs) so reviewers can follow how the query got its final shape. Description-only; no code touched.

**Comment by GonzaloSecades** [comment]
## Review cycle — round 3

**Addressed (1/1):** [low] PR description still described the superseded per-column `unnest WITH ORDINALITY` bulk-create design — updated to the shipped `jsonb_to_recordset` shape, plus a Review history section tracing the design evolution across all rounds. Description-only change; no code, no new commit.

**Rebutted:** none. This was Copilot's only round-3 finding; the code itself came back clean.

3 of 3 automated iterations used — any further Copilot rounds are Gonzalo-triggered per the cap.

**Comment by GonzaloSecades** [comment]
## Registry restructure + 0006 relocation

Per Gonzalo's direction, `post-migration-improvements/` is now split into agent-owned subdirectories: `claude-backend-improvements/` (Claude's numbered parity-deviation entries — 0001–0005 moved on master, **0006 moved on this branch**) and `codex-backend-improvements/` (Codex's cross-cutting optimization program). The root README keeps the rule/template and now carries a Layout section and the full index including 0006; CLAUDE.md names the entry location and keeps each agent out of the other's directory. Master merged into this branch (only conflict: the README index — resolved by taking master's restructured version, which already lists 0006 at its new path). graphify refreshed on the merged tree.

Note: this restructure merge landed after the 3-iteration cap was reached; it contains no query/code changes (docs moves + generated artifacts only). Awaiting CI, then merge on Gonzalo's word (already given in session).
